In [2]:
import pandas as pd
import requests
import io
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [3]:
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab

In [4]:
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab

### Webscarping to Extract Toronto neighborhood data from a Wikipedia page

In [5]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
# get the contents of the webpage in text format and store in a variable called data
html_data  = requests.get(url).text
soup = BeautifulSoup(html_data,"html5lib")

In [6]:
#read the data
soup

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en"><head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"540cf718-c4ea-4278-99a9-752fb1130fb4","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":1013111980,"wgRevisionId":1013111980,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Wikipedia semi-protected pag

**Step by Step Instructions:**

**After retreiving the URL and creating a Beautiful soup object** 

 **Firstly create a list**  

 **Later after finding the table and table data  create a dictionary called cell having 3 keys PostalCode, Borough and Neighborhood.**

**As postal code contains upto 3 characters extract that using tablerow.p.text**

 **Next use split ,strip and replace functions for getting Borough and Neighborhood information.**.

 **Append to the list**  

 **Create a dataframe with list**

In [7]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [8]:
df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

### Data Cleaning

In [9]:
#checking NAN values
df.isnull().T.any().T.sum()

0

In [10]:
df['Neighborhood'].isnull().values.any()

False

### Data Visualization

In [11]:
df.shape

(103, 3)

The dataframe has 103 rows and 3 columns

### PART-2 
### Getting the latitude and the longitude coordinates of each neighborhood

In [12]:
url="https://cocl.us/Geospatial_data"
s=requests.get(url).content
lat_lon=pd.read_csv(io.StringIO(s.decode('utf-8')))
lat_lon

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750071 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

### Concatinate latitude and longitute with Neughborhood dataframe

In [13]:
neighborhoods = df.join(lat_lon.set_index('Postal Code'), on='PostalCode')
neighborhoods.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

### Part -3 
### Explore and cluster the neighborhoods in Toronto

In [14]:
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate for Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate for Toronto are 43.6534817, -79.3839347.


### Create a map of New York with neighborhoods superimposed on top

In [15]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [16]:
CLIENT_ID = '5RVOEB2NPJGTPJCRTC0EAHHKIDXE1JDXWZ0TTNEAV5AH1HE5' # your Foursquare ID
CLIENT_SECRET = '3PZBD2502RAPDIGI4LG5AZBITK254MEA2WN3TUFISVETM3AD' # your Foursquare Secret
#ACCESS_TOKEN = 'EPTLQ1JFAOBXP0IFVJQWBLWWRBAGH4HTXI5BMGMCXKIEVJ42' # your FourSquare Access Token
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5RVOEB2NPJGTPJCRTC0EAHHKIDXE1JDXWZ0TTNEAV5AH1HE5
CLIENT_SECRET:3PZBD2502RAPDIGI4LG5AZBITK254MEA2WN3TUFISVETM3AD


### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [17]:
neighborhoods.loc[0, 'Neighborhood']

'Parkwoods'

Get the neighborhood's latitude and longitude values.

In [18]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

In [19]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=5RVOEB2NPJGTPJCRTC0EAHHKIDXE1JDXWZ0TTNEAV5AH1HE5&client_secret=3PZBD2502RAPDIGI4LG5AZBITK254MEA2WN3TUFISVETM3AD&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [20]:
#Send the GET request and examine the results
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '606f95d8e649ec7f7c2bf630'},
 'response': {'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'distance': 245,
        'cc': 'CA'

In [21]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [22]:
#clean the json and structure it into a pandas dataframe

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:5: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name                  categories        lat  \
0                    Brookbanks Park                        Park  43.751976   
1                      Variety Store           Food & Drink Shop  43.751974   
2  Corrosion Service Company Limited  Construction & Landscaping  43.752432   
3                      GreenWin pool                        Pool  43.756232   

         lng  
0 -79.332140  
1 -79.333114  
2 -79.334661  
3 -79.333842

In [23]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


## Explore Neighborhoods in Manhattan


#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [24]:
neighborhoods

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

Selecting the boroughs that contain the word Toronto

In [25]:
toronto_df = neighborhoods[neighborhoods['Borough'].str.contains('Toronto', regex=False, case=False, na=False)]
toronto_df

PostalCode                 Borough  \
2          M5A        Downtown Toronto   
9          M5B        Downtown Toronto   
15         M5C        Downtown Toronto   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
54         M4M            East Toronto   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
96         M4X        Downtown Toronto   
97         M5X        Downtown Toronto   
99         M4Y        Downtown Toronto   
100        M7Y   East Toronto Business   

                                          Neighborhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
35                                  The Danforth  East  43.685347 -79.338106  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68                            Forest Hill North & West  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                                  North Toronto West  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                              

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        #print(url)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
        #print(venues_list)

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [27]:
# type your answer here
toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

In [29]:
#Let's check the size of the resulting dataframe
print(toronto_venues.shape)
toronto_venues.head()

(1565, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0           Tandem Coffee       43.653559       -79.361809   
1        Roselle Desserts       43.653447       -79.362017   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0          Coffee Shop  
1               Bakery  
2  Distribution Center  
3           Restaurant  
4                  Spa

### Let's check how many venues were returned for each neighborhood

In [30]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            60   
Brockton, Parkdale Village, Exhibition Place                           25   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   71   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                       10   
Dufferin, Dovercourt Village                                           15   
Enclave of M4L                                                         15   
Enclave of M5E                                                         99   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              64   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     18   
Parkdale, Roncesvalles                                                 15   
Regent Park, Harbourfront                                              43   
Richmond, Adelaide, King                                               93   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     30   
St. James Town                                                         82   
St. James Town, Cabbagetown                                            43   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    20   
The Beaches                                                             4   
The Danforth  East                                                      4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             60   
Brockton, Parkdale Village, Exhibition Place                            25   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    71   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                        10   
Dufferin, Dovercourt Village        

#### Let's find out how many unique categories can be curated from all the returned venues

In [31]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 230 uniques categories.


## Analyze Each Neighborhood

In [51]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0            0        0                   0               0                0   
1            0        0                   0               0                0   
2            0        0                   0               0                0   
3            0        0                   0               0                0   
4            0        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       1     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  Cheese Shop  \
0                          0            0                     0            0   
1                          0            0                     0            0   
2                          0            0                     0            0   
3                          0     

In [52]:
toronto_onehot.shape

(1565, 230)

In [53]:
#Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.040000  0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.066667   
3                                  Central Bay Street     0.016129  0.000000   
4                                            Christie     0.000000  0.000000   
5                                Church and Wellesley     0.028169  0.000000   
6                      Commerce Court, Victoria Hotel     0.000000  0.000000   
7                                          Davisville     0.000000  0.000000   
8                                    Davisville North     0.000000  0.000000   
9                        Dufferin, Dovercourt Village     0.000000  0.000000   
10                                     Enclave of M4L     0.000000  0.000000   
11                                     Enclave of M5E     0.010101  0.000000   
12             First Canadian Place, Underground city     0.000000  0.000000   
13                           Forest Hill North & West     0.000000  0.000000   
14                           Garden District, Ryerson     0.000000  0.000000   
15  Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
16                      High Park, The Junction South     0.000000  0.000000   
17                     India Bazaar, The Beaches West     0.000000  0.000000   
18          Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
19                                      Lawrence Park     0.000000  0.000000   
20                           Little Portugal, Trinity     0.022727  0.000000   
21                        Moore Park, Summerhill East     0.000000  0.000000   
22                                 North Toronto West     0.055556  0.000000   
23                             Parkdale, Roncesvalles     0.000000  0.000000   
24                          Regent Park, Harbourfront     0.023256  0.000000   
25                           Richmond, Adelaide, King     0.000000  0.000000   
26                                           Rosedale     0.000000  0.000000   
27                                           Roselawn     0.000000  0.000000   
28                                 Runnymede, Swansea     0.033333  0.000000   
29                                     St. James Town     0.000000  0.000000   
30                        St. James Town, Cabbagetown     0.000000  0.000000   
31                                    Studio District     0.027027  0.000000   
32  Summerhill West, Rathnelly, South Hill, Forest...     0.000000  0.000000   
33                The Annex, North Midtown, Yorkville     0.000000  0.000000   
34                                        The Beaches     0.000000  0.000000   
35                                 The Danforth  East     0.000000  0.000000   
36                       The Danforth West, Riverdale     0.023810  0.000000   
37           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
38                     University of Toronto, Harbord     0.030303  0.000000   

    Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0             0.000000        0.000000         0.000000          0.000000   
1             0.000000        0.000000         0.000000          0.000000   
2             0.066667        0.133333         0.133333          0.133333   
3             0.000000        0.000000         0.000000          0.000000   
4             0.000000        0.000000         0.000000          0.000000   
5             0.000000        0.000000         0.000000          0.000000   
6             0.000000        0.000000         0.000000          0.000000   
7             0.000000        0.000000         0.000000          0.000000   
8             0.000000        0.000000         0.000000          0.000000   
9             0.000000        0.000000         0.000000          0.000

In [54]:
toronto_grouped.shape

(39, 230)

### Let's print each neighborhood along with the top 5 most common venues


In [55]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.08
1          Bakery  0.05
2    Cocktail Bar  0.05
3  Farmers Market  0.03
4        Beer Bar  0.03


----Brockton, Parkdale Village, Exhibition Place----
                   venue  freq
0                   Café  0.12
1         Breakfast Spot  0.08
2  Performing Arts Venue  0.08
3            Coffee Shop  0.08
4                 Bakery  0.08


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3       Coffee Shop  0.07
4           Airport  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1      Sandwich Place  0.06
2                Café  0.05
3  Italian Restaurant  0.05
4     Thai Restaurant  0.03


----Christie----
           venue  freq
0  Grocery Store  0.24
1           Café  0.18
2           Park  0.12
3     Baby Sto

Let's put that into a pandas dataframe
First, let's write a function to sort the venues in descending order.

In [56]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [57]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                Bakery           Cocktail Bar        Farmers Market   
1        Breakfast Spot  Performing Arts Venue           Coffee Shop   
2       Airport Service       Airport Terminal           Coffee Shop   
3        Sandwich Place                   Café    Italian Restaurant   
4                  Café                   Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0              Beer Bar              Pharmacy             Cheese Shop   
1                Bakery           Yoga Studio  Furniture / Home Store   
2               Airport      Sculpture Garden     Rental Car Location   
3       Thai Restaurant       Bubble Tea Shop            Burger Joint   
4           Coffee Shop             Nightclub      Italian Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant    Seafood Restaurant      Indian Restaurant  
1         Burrito Place          Climbing Gym      Convenience Store  
2         Boat or Ferry              Boutique                  Plane  
3           Salad Place                   Spa      French Restaurant  
4           Candy Store    Athletics & Sports             Restaurant

In [58]:
toronto_grouped.head()

Neighborhood  Yoga Studio   Airport  \
0                                        Berczy Park     0.000000  0.000000   
1       Brockton, Parkdale Village, Exhibition Place     0.040000  0.000000   
2  CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.066667   
3                                 Central Bay Street     0.016129  0.000000   
4                                           Christie     0.000000  0.000000   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0            0.000000        0.000000         0.000000          0.000000   
1            0.000000        0.000000         0.000000          0.000000   
2            0.066667        0.133333         0.133333          0.133333   
3            0.000000        0.000000         0.000000          0.000000   
4            0.000000        0.000000         0.000000          0.000000   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                  0.0           0.0       0.0     0.016667    0.000000   
1                  0.0           0.0       0.0     0.000000    0.000000   
2                  0.0           0.0       0.0     0.000000    0.000000   
3                  0.0           0.0       0.0     0.000000    0.016129   
4                  0.0           0.0       0.0     0.000000    0.000000   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                  0.0               0.0            0.000000            0.0   
1                  0.0               0.0            0.000000            0.0   
2                  0.0               0.0            0.000000            0.0   
3                  0.0               0.0            0.000000            0.0   
4                  0.0               0.0            0.058824            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery      Bank   Bar  \
0        0.0    0.000000    0.016667    0.05  0.000000  0.00   
1        0.0    0.000000    0.000000    0.08  0.000000  0.04   
2        0.0    0.000000    0.000000    0.00  0.000000  0.00   
3        0.0    0.000000    0.000000    0.00  0.016129  0.00   
4        0.0    0.058824    0.000000    0.00  0.058824  0.00   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.016667  0.016667              0.0  0.033333   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  \
0         0.0            0.016667  0.016667       0.000000   0.000000   
1         0.0            0.000000  0.000000       0.000000   0.000000   
2         0.0            0.000000  0.000000       0.066667   0.000000   
3         0.0            0.000000  0.000000       0.000000   0.016129   
4         0.0            0.000000  0.000000       0.000000   0.000000   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0  0.000000                   0.0        0.016667      0.0         0.000000   
1  0.000000                   0.0        0.080000      0.0         0.000000   
2  0.066667                   0.0        0.000000      0.0         0.000000   
3  0.000000                   0.0        0.000000      0.0         0.032258   
4  0.000000                   0.0        0.000000      0.0         0.000000   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
0       0.0      0.000000           0.00       0.0      0.0  0.016667   
1       0.0      0.000000           0.04       0.0      0.0  0.120000   
2       0.0      0.000000           0.00       0.0      0.0  0.000000   
3       0.0      0.032258           0.00       0.0      0.0  0.048387   
4       0.0      0.000000           0

In [59]:
neighborhoods_venues_sorted.shape

(39, 11)

# Cluster Neighborhoods


Run k-means to cluster the neighborhood into 5 clusters.

In [60]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [61]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough               Neighborhood   Latitude  \
2         M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
9         M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
15        M5C  Downtown Toronto             St. James Town  43.651494   
19        M4E      East Toronto                The Beaches  43.676357   
20        M5E  Downtown Toronto                Berczy Park  43.644771   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2  -79.360636               0           Coffee Shop                Bakery   
9  -79.378937               0        Clothing Store           Coffee Shop   
15 -79.375418               0                  Café           Coffee Shop   
19 -79.293031               0     Health Food Store                 Trail   
20 -79.373306               0           Coffee Shop                Bakery   

        3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2                        Park                   Pub        Breakfast Spot   
9   Middle Eastern Restaurant       Bubble Tea Shop        Cosmetics Shop   
15               Cocktail Bar        Cosmetics Shop   Moroccan Restaurant   
19                        Pub           Yoga Studio                Museum   
20               Cocktail Bar        Farmers Market              Beer Bar   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2                       Café               Theater        Cosmetics Shop   
9                       Café        Lingerie Store  Fast Food Restaurant   
15        Seafood Restaurant              Creperie            Restaurant   
19  Mediterranean Restaurant           Men's Store         Metro Station   
20                  Pharmacy           Cheese Shop            Restaurant   

   9th Most Common Venue     10th Most Common Venue  
2                    Spa                 Shoe Store  
9              Bookstore                    Theater  
15    Italian Restaurant           Department Store  
19    Mexican Restaurant  Middle Eastern Restaurant  
20    Seafood Restaurant          Indian Restaurant

### Finally, let's visualize the resulting clusters

In [62]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

Cluster 1

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2          Downtown Toronto               0           Coffee Shop   
9          Downtown Toronto               0        Clothing Store   
15         Downtown Toronto               0                  Café   
19             East Toronto               0     Health Food Store   
20         Downtown Toronto               0           Coffee Shop   
24         Downtown Toronto               0           Coffee Shop   
25         Downtown Toronto               0         Grocery Store   
30         Downtown Toronto               0           Coffee Shop   
31             West Toronto               0              Pharmacy   
36         Downtown Toronto               0           Coffee Shop   
37             West Toronto               0                   Bar   
41             East Toronto               0      Greek Restaurant   
42         Downtown Toronto               0           Coffee Shop   
43             West Toronto               0                  Café   
47             East Toronto               0                  Park   
48         Downtown Toronto               0           Coffee Shop   
54             East Toronto               0           Coffee Shop   
67          Central Toronto               0                   Gym   
69             West Toronto               0                  Café   
73          Central Toronto               0           Coffee Shop   
74          Central Toronto               0                  Café   
75             West Toronto               0        Breakfast Spot   
79          Central Toronto               0           Pizza Place   
80         Downtown Toronto               0                  Café   
81             West Toronto               0                  Café   
84         Downtown Toronto               0                  Café   
86          Central Toronto               0           Coffee Shop   
87         Downtown Toronto               0        Airport Lounge   
92   Downtown Toronto Stn A               0           Coffee Shop   
96         Downtown Toronto               0           Coffee Shop   
97         Downtown Toronto               0           Coffee Shop   
99         Downtown Toronto               0           Coffee Shop   
100   East Toronto Business               0    Light Rail Station   

    2nd Most Common Venue      3rd Most Common Venue  \
2                  Bakery                       Park   
9             Coffee Shop  Middle Eastern Restaurant   
15            Coffee Shop               Cocktail Bar   
19                  Trail                        Pub   
20                 Bakery               Cocktail Bar   
24         Sandwich Place                       Café   
25                   Café                       Park   
30                   Café                 Restaurant   
31                 Bakery                        Bar   
36               Aquarium                       Café   
37            Coffee Shop      Vietnamese Restaurant   
41     Italian Restaurant                Coffee Shop   
42                  Hotel                       Café   
43         Breakfast Spot      Performing Arts Venue   
47   Fast Food Restaurant                        Gym   
48             Restaurant                      Hotel   
54                Brewery                       Café   
67                  Hotel          Convenience Store   
69     Mexican Restaurant            Thai Restaurant   
73         Clothing Store                Yoga Studio   
74         Sandwich Place                Coffee Shop   
75              Gift Shop                Coffee Shop   
79           Dessert Shop             Sandwich Place   
80         Sandwich Place                        Bar   
81       Sushi Restaurant                Coffee Shop   
84            Coffee Shop         Mexican Restaurant   
86           Liquor Store                 Bagel Shop   
87        Airport Service           Airport Terminal   
92     Seafood Restaurant                 Restaur

Cluster 2

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue     4th Most Common Venue  \
62                Museum        Massage Studio  Mediterranean Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62           Men's Store         Metro Station    Mexican Restaurant   

        8th Most Common Venue 9th Most Common Venue  \
62  Middle Eastern Restaurant    Miscellaneous Shop   

        10th Most Common Venue  
62  Modern European Restaurant

Cluster 3

In [65]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
35  East York/East Toronto               2                  Park   
68         Central Toronto               2                  Park   
91        Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
35     Convenience Store         Metro Station           Yoga Studio   
68                 Trail         Jewelry Store      Sushi Restaurant   
91            Playground                 Trail           Yoga Studio   

   5th Most Common Venue     6th Most Common Venue     7th Most Common Venue  \
35                Museum  Mediterranean Restaurant               Men's Store   
68           Yoga Studio             Movie Theater  Mediterranean Restaurant   
91                Museum  Mediterranean Restaurant               Men's Store   

   8th Most Common Venue      9th Most Common Venue     10th Most Common Venue  
35    Mexican Restaurant  Middle Eastern Restaurant         Miscellaneous Shop  
68           Men's Store              Metro Station         Mexican Restaurant  
91         Metro Station         Mexican Restaurant  Middle Eastern Restaurant

Cluster 4

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
61  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
61              Bus Line           Swim School           Yoga Studio   

   5th Most Common Venue     6th Most Common Venue 7th Most Common Venue  \
61         Movie Theater  Mediterranean Restaurant           Men's Store   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
61         Metro Station    Mexican Restaurant  Middle Eastern Restaurant

Cluster 5

In [67]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue     4th Most Common Venue  \
83           Yoga Studio                Museum  Mediterranean Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
83           Men's Store         Metro Station    Mexican Restaurant   

        8th Most Common Venue 9th Most Common Venue  \
83  Middle Eastern Restaurant    Miscellaneous Shop   

        10th Most Common Venue  
83  Modern European Restaurant